### DISCUSSION TITLE
HttpRequestMessage for HttpClient methods

### DISCUSSION THREAD

**Discussion by Raphsodyz** [richardszalay/mockhttp#137] [Q&A] [answered]
Hey guys!

I'm have a method in my business layer that i'm want to test it. I'm using the package to throw a TooManyRequest result something like:

`var mockHttp = new MockHttpMessageHandler();`
`mockHttp.When("some url").Respond(_ => { return new HttpResponseMessage(HttpStatusCode.TooManyRequests);});`

When i'm call the method, i'm want that result of the request be a TooManyRequests return. Specifially, i'm trying to do a test case for my if condition in my unit test: 

`HttpResponseMessage response = await client.GetAsync("some url");`
`if (!response.IsSuccessStatusCode) throw new HttpRequestException("some exception message");`

How to do that? The first code don't work. Glad for a help!

**Comment by diegodrf** [comment]
Hi @Raphsodyz .

I ran your code and it's working properly. When we call the mocked url the return is 429 (Too many requests).

Verify if the URL passed to the `When()` method has a perfect match.

You said it's not working, but didn't show the error message. Could you complete the example with the error message?

**Reply by Raphsodyz** [reply]
Hey @diegodrf ! Thanks for the answer!

I'm not sure what's happening, but, when i'm debug my test passing through the code it returns the TooManyRequest as expected, but when i'm run the test without debugging it, it's return a null result in the exception. 😂

My test case is:

``` C#
        [Fact]
        public async void LaunchApiBusiness_UpdateLaunch_TooManyRequestError()
        {
            //Arrange
            var mockHttp = new MockHttpMessageHandler();
            mockHttp.When($"{EndPoints.TheSpaceDevsLaunchEndPoint}*")
                .Respond(_ => { return new HttpResponseMessage(HttpStatusCode.TooManyRequests);});

            var launchApiBusiness = new Mock<ILaunchApiBusiness>();
            var launchRepository = new Mock<ILaunchRepository>();
            var launchViewRepository = new Mock<ILaunchViewRepository>();
            var client = mockHttp.ToHttpClient();

            var uow = new Mock<IUnitOfWork>();
            var factoryClient = new Mock<IHttpClientFactory>();
            var mapper = new Mock<IMapper>();

            uow.Setup(u => u.Repository(typeof(ILaunchRepository))).Returns(launchRepository.Object);
            factoryClient.Setup(f => f.CreateClient(It.IsAny<string>())).Returns(() => client);
            launchRepository.Setup(l => l.Get(It.IsAny<Expression<Func<Launch, bool>>>(), string.Empty)).ReturnsAsync(TestLaunchObjects.Test1());
            launchRepository.Setup(l => l.GetTransaction()).ReturnsAsync(Mock.Of<IDbContextTransaction>());

            var response = await client.GetAsync($"{EndPoints.TheSpaceDevsLaunchEndPoint}{TestLaunchDTOObjects.Test1().Id}");
            var business = new LaunchApiBusiness(uow.Object, factoryClient.Object, mapper.Object);

            //Act
            var result = business.UpdateLaunch(TestLaunchObjects.Test1().Id);

            //Assert
            Assert.NotNull(result);
            Assert.IsType<AggregateException>(result.Exception);
            Assert.Equal(result.Exception.Message, "One or more errors occurred. (TooManyRequests - Attention! The SpaceDevs API endpoint returned an error.)");
        }
```

And my controller is:

``` C#
public async Task<LaunchView> UpdateLaunch(Guid? launchId)
        {
            ILaunchBusiness _launchBusiness = GetBusiness(typeof(ILaunchBusiness)) as ILaunchBusiness;

            if (launchId == null)
                throw new ArgumentNullException(ErrorMessages.NullArgument);

            Expression<Func<Launch, bool>> launchQuery = l => l.Id == launchId && l.EntityStatus == EStatus.PUBLISHED.GetDisplayName();
            var launch = await _launchBusiness.Get(filter: launchQuery) ?? throw new KeyNotFoundException(ErrorMessages.KeyNotFound);

            using var trans = await _repository.GetTransaction();
            var client = _client.CreateClient();
            try
            {
                string url = $"{EndPoints.TheSpaceDevsLaunchEndPoint}{launch.ApiGuId}";
                HttpResponseMessage response = await client.GetAsync(url);
                if (!response.IsSuccessStatusCode)
                    throw new HttpRequestException($"{response.StatusCode} - {ErrorMessages.LaunchApiEndPointError}");

                var updatedLaunch = (await response.Content.ReadFromJsonAsync<LaunchDTO>() ?? throw new HttpRequestException(ErrorMessages.DeserializingEndPointContentError)) ?? throw new KeyNotFoundException(ErrorMessages.KeyNotFound);
                launch = _mapper.Map<Launch>(updatedLaunch);
                launch.EntityStatus = EStatus.PUBLISHED.GetDisplayName();

                await trans.CommitAsync();

                ILaunchViewBusiness _launchViewBusiness = GetBusiness(typeof(ILaunchViewBusiness)) as ILaunchViewBusiness;
                await _launchViewBusiness.RefreshView();

                var result = await _launchViewBusiness.GetById(l => l.Id == launchId && l.EntityStatus == EStatus.PUBLISHED.GetDisplayName());                
                return result;
            }
            catch (HttpRequestException ex)
            {
                throw ex;
            }
            catch (KeyNotFoundException ex)
            {
                throw ex;
            }
            catch (Exception ex)
            {
                await trans.RollbackAsync();
                throw ex;
            }
            finally
            {
                await trans.DisposeAsync();
            }
        }
```

Now I think this problem has nothing to do with the package, but another thing. haha

Without debbuging:

![image](https://github.com/richardszalay/mockhttp/assets/22720154/b2acf983-5fec-4099-8acd-aa849abeddee)

Debugging:

![image](https://github.com/richardszalay/mockhttp/assets/22720154/ba9809a3-4e25-4f1c-9936-75f9f11c37a6)
![image](https://github.com/richardszalay/mockhttp/assets/22720154/d6d0a115-13d1-4f6a-b8ac-d89dcca3d181)

So strange!

**Reply by diegodrf** [reply, marked as answer]
I think the error is in your code.

You are trying to read the property `LaunchView.Exception` in `Assert.IsType<AggregateException>(result.Exception);` but you are not saving the value of the exception in this property and returning it gracefully. 

When the status code is not 200 you throw an exception and not catching it.
```cs
if (!response.IsSuccessStatusCode)
                    throw new HttpRequestException($"{response.StatusCode} - {ErrorMessages.LaunchApiEndPointError}");
```

The correct way to test it is by calling `Assert.ThrowsAsync` if you are using XUnit or equivalent in another package.

Pay attention to await too. You are using an async method and not awaiting. So at line `var result = business.UpdateLaunch(TestLaunchObjects.Test1().Id);` in your Act you "fire and forget" the function and the value of `result` will be `Task` and not the expected return.

**Reply by Raphsodyz** [reply]
Adjusting the code to correctly await the result and catch the exception by Assert.ThrowsAsync works as a expected!!
Thanks so much @diegodrf 😁
